sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java

```diff
@@ -149,6 +150,29 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
         return Channels.newChannel(
             new BZip2CompressorInputStream(Channels.newInputStream(channel)));
       }
+    },
+
+    /**
+     * Reads a byte channel assuming it is compressed with zip.
+     */
+    ZIP {
+      @Override
+      public boolean matches(String fileName) {
+        return fileName.toLowerCase().endsWith(".zip");
+      }
+
+      public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channel)
+        throws IOException {
+        ZipInputStream zip = new ZipInputStream(Channels.newInputStream(channel));
+        if (zip.getNextEntry() != null) {
+          if (zip.getNextEntry() != null) {
+            throw new IllegalArgumentException("Multi-entries ZIP file is not supported");
+          } else {
+            return Channels.newChannel(zip);
```

Comment:
does this work? Based on how you wrote the code, I would expect that `zip.getNextEntry()` actually mutates the ZipInputStream in some way. Is it still valid to open a stream from the zip after calling getNextEntry twice?